python_modules/libraries/dagster-dg/dagster_dg/component.py

```diff
@@ -24,6 +25,30 @@ def key(self) -> str:
         return self.name
 
 
+class LocalComponentTypes:
+    @classmethod
+    def from_dg_context(cls, dg_context: "DgContext", component_dirs: Sequence[Path]):
+        # TODO: cache
+
+        raw_local_component_data = dg_context.external_components_command(
+            ["list", "local-component-types", *component_dirs]
+        )
+        local_component_data = json.loads(raw_local_component_data)
+
+        components_by_path = {str(path): {} for path in component_dirs}
+        for entry in local_component_data:
+            components_by_path[entry["directory"]][entry["key"]] = RemoteComponentType(
+                **entry["metadata"]
+            )
+        return cls(components_by_path)
+
+    def __init__(self, components_by_path: dict[str, dict[str, RemoteComponentType]]):
+        self._components_by_path = copy.copy(components_by_path)
+
+    def get(self, path: Path, key: str) -> RemoteComponentType:
+        return self._components_by_path[str(path)][key]
```

Comment:
I feel like the name of this class is pretty confusing -- it's called LocalComponentTypes but it's storing a bunch of RemoteComponentTypes

It does sort of feel like we should just have a param for RemoteComponentRegistry to load additional local component types. Like from_dg_context having an optional additional_component_dirs parameter.

That would simplify code that uses this stuff too -- you don't have to juggle two separate objects around and check which one you should be using

Reply:
Yeah I agree the name sucks. I do think it's somewhat useful that it's separate in that
1. it's not comprehensive, e.g. a single instance represents a set of local component types from a few components, not necessarily the whole code location
2. you may only want to load it in some scenarios, in others you only want to load the global `RemoteComponentRegustry` component type list

>  optional `additional_component_dirs` 

This would work, but would need a refactor of `RemoteComponentRegustry` to no longer use component type as a unique key, since local components can have name overlap (since they're 'scoped' to their folder)

Reply:
i see what you mean about uniqueness -- I guess just more fundamentally I think "bag of remote components that we want to do stuff with" should be a single object. [this code](https://github.com/dagster-io/dagster/pull/27214/files#diff-9469a20ba4790f64fbde6ba6c4e6dbc14a1ab3c58837517f2a60c1b5b9d27f9bR322-R334) feels pretty brittle and painful to work with as time goes on.

what other cases are you imagining where we'd want only the LocalComponents? 

i could be convinced that we could stick with this sort of thing (ideally with a rename) until a new usecase comes along for us to refactor based off of, just feels like this pattern shouldn't stay around forever

might be a matter of subclassing or something

Reply:
yeah let me feel out what it would look like to rewrite the existing remote components class

Reply:
did a restructure, I'm happier with this approach